src/core/fieldformatter/qgscheckboxfieldformatter.cpp

```diff
@@ -0,0 +1,47 @@
+/***************************************************************************
+  qgscheckboxfieldformatter.cpp - QgsCheckBoxFieldFormatter
+
+ ---------------------
+ begin                : 23.09.2019
+ copyright            : (C) 2019 by Denis Rouzaud
+ email                : denis@opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QObject>
+
+#include "qgscheckboxfieldformatter.h"
+#include "qgsapplication.h"
+
+
+QString QgsCheckBoxFieldFormatter::id() const
+{
+  return QStringLiteral( "CheckBox" );
+}
+
+QString QgsCheckBoxFieldFormatter::representValue( QgsVectorLayer *layer, int fieldIndex, const QVariantMap &config, const QVariant &cache, const QVariant &value ) const
+{
+  Q_UNUSED( layer )
+  Q_UNUSED( fieldIndex )
+  Q_UNUSED( cache )
+  Q_UNUSED( config )
+
+
+  if ( value.isNull() || !value.canConvert<bool>() )
```

Comment:
`!value.canConvert<bool>()`

Is this even possible, or is every qvariant convertible to a bool?

But -- more importantly -- shouldn't this formatter also consider the true/false values which can be set for a non-boolean field using the checkbox wrapper? Or should the raw values for non-boolean variant types be used instead (I think this is what I'd prefer).

Reply:
> Is this even possible, or is every qvariant convertible to a bool?
good question. 

> But -- more importantly -- shouldn't this formatter also consider the true/false values which can be set for a non-boolean field using the checkbox wrapper? Or should the raw values for non-boolean variant types be used instead (I think this is what I'd prefer).

That's really correct this has to be handled as well, i.e. compare the values to the true/false configuration.

I am really not sure we should display the raw values. Let's consider that there is no boolean in SQLite for instance. I'd prefer have true/false than 0/1 written.
It would be more interesting to have configurable represent value but that's maybe over-engineering the problem and out-of-scope here.

Reply:
I also tend towards true/false(/NULL)
Other values should only ever be shown if there's a problem converting them to a boolean (let's say the configuration is for `t` and `f` and a `p` is in the field, then it show something like `(p)`)

Reply:
I like the idea of `(p)`. I'm definitely NOT a fan of silently mapping these across to NULL -- that's misleading.

Reply:
ok this is done now.

```
if field type is bool:
  NULL => nullRepresentation
  true => tr("true")
  false => tr("false")
else
  if cannot convert to string (like json integer list) => (invalid)
  if == checkedstate => tr("true")
  if == uncheckedstate => tr("false")
  else (value.toString)
```